src/nvim/message.c

```diff
@@ -226,6 +227,11 @@ bool msg_attr_keep(char_u *s, int attr, bool keep, bool multiline)
     add_msg_hist((const char *)s, -1, attr, multiline);
   }
 
+  if ((State & CMDLINE) && !emsg_on_display && !redraw_cmdline) {
```

Comment:
This seems quite ad-hoc. What problems does it solve? What happens if `cmdheight` or `ext_cmdline` is used, so that there is ample space to display both a message and cmdline without scroll?

Reply:
The logic has been changed slightly, but the gist is that if we are in the cmdline, don't print any non-error messages.

For error messages, the caller has to call `redrawcmdline` after printing the error message to correctly draw the cmdline.

Reply:
Ah, I see what you mean - it is possible for an external UI to handle both cmdline and messages separately (unlike in the TUI where they are quite tightly coupled).